Activity dao

wszarmach opened the pull request:
First phase to move to activity dao. Modified code to make use of dao get operations only.

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityDao.java:
```diff
@@ -0,0 +1,44 @@
+            @Override
+            public List<Key> execute() {
+                return Datastore.query(meta, organization.getId()).asKeyList();
+            }
```
This is an ancestor query, to make it so that it works with transactions, you should write it as
```java
  Datastore.query(Datastore.getCurrentTransaction(), meta, organization.getId()).asKeyList();
```

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityPackageActivityDao.java:
```diff
@@ -0,0 +1,49 @@
+    public List<ActivityPackageActivity> getBy(final Activity activity) {
+        ActivityPackageActivityMeta meta = getMeta();
+        try {
+            ActivityType activityType = activityTypeDao.get(activity.getActivityTypeRef().getKey());
```
1. Warning:  `activity.getActivityTypeRef().getKey()` might be **null**
2. Warning: if someone deleted the activityType, then `activityTypeDao.get(...)` will throw EntityNotFoundException

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityPackageActivityDao.java:
```diff
@@ -0,0 +1,49 @@
+        ActivityPackageActivityMeta meta = getMeta();
+        try {
+            ActivityType activityType = activityTypeDao.get(activity.getActivityTypeRef().getKey());
+            final Key organisationKey = activityType.getOrganizationRef().getKey();
```
Here also, `activityType.getOrganizationRef().getKey();` might be **null**.  Maybe in both cases, if the key is null we should throw not found?

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityPackageActivityDao.java:
```diff
@@ -0,0 +1,49 @@
+                            .filter(new CompositeCriterion(meta,
+                                    Query.CompositeFilterOperator.AND,
+                                    meta.activityRef.equal(activity.getId()))).asKeyList();
+                }
```
1. Ancestor query should be transactional, if you call this method with an active transaction, the `activityTypeDao.get` will be in the transaction, but the query won't.
 2. I don't understand why you need composite criterion here, what are you "and"ing?

This could be written as:
```java
Datastore
   .query(Datastore.getCurrentTransaction(), meta, organizationKey)
   .filter(meta.activityRef.equal(activity.getId())
   .asKeyList();
```

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityPackageActivityDao.java:
```diff
@@ -0,0 +1,49 @@
+                }
+            });
+        } catch (EntityNotFoundException e) {
+            // Is this possible?
```
I would not silently swallow this exception but rather have the method query throw it.
```java
  public List<ActivityPackageActivity> getBy(final Activity activity) throws EntityNotFoundException
```

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/dao/common/ActivityTypeDao.java:
```diff
@@ -0,0 +1,51 @@
+            @Override
+            public List<Key> execute() {
+                return Datastore.query(meta, organization.getId()).asKeyList();
+            }
```
* Ancestor query... `getCurrentTransaction()`

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/activity/ActivityType.java:
```diff
@@ -79,7 +80,7 @@ public String getName() {
     public void setName(String name) {
         this.name = name;
-        this.lcName = name;
+        setLcName(StringUtils.lowerCase(name));
```
the lowerCase here is no harm, but as an FYI we use `LowerCaseListener` so when it is persisted it get lowercased...

```java
    @Attribute(listener = LowerCaseListener.class)
    private String lcName;
```

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java:
```diff
@@ -64,12 +64,18 @@
-            return ActivityServiceFactory.getActivityService().getActivityTypes(organization);
+            Organization organization = organizationDao.get(organizationId);
+            return activityTypeDao.getBy(organization);
         } catch (EntityNotFoundException e) {
```
This I don't like.  Why do we need to get the organization if all we need in the getBy query is the organizationId?

I'm all for a `activityTypeDao.getByOrganizationId(organizationId);`, you dislike it?

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java:
```diff
@@ -189,17 +190,19 @@ public boolean apply(@Nullable Activity input) {
-                all.addAll(ActivityServiceFactory.getActivityService().getActivities(organization));
+                Organization organization = organizationDao.get(organizationId);
+                all.addAll(activityDao.getBy(organization));
             }
```
same here, no real need to get organization

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/OrgMemberChecker.java:
```diff
@@ -39,8 +46,8 @@ protected OrgMemberChecker initialValue() {
-                    ActivityType activityType = activity.getActivityTypeRef().getModel();
+                    Activity activity = getActivity(id);
+                    ActivityType activityType = getActivityType(activity.getActivityTypeRef().getKey());
                     return getOrganization(activityType.getOrganizationRef().getKey());
```
NullPointer ?  `if(activity==null) return null;`

Same for all getActivity, getActivityType, etc

krico commented on schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/transform/JasActivityTransformer.java:
```diff
@@ -20,7 +25,8 @@ public JasActivityTransformer() {
         BeanUtil.copyProperties(external, internal);
-        ActivityType activityType = internal.getActivityTypeRef().getModel();
+        internal.getActivityTypeRef().getKey();
+        ActivityType activityType = getActivityType(internal);
```
What's this line for `internal.getActivityTypeRef().getKey();`

krico commented:
Was able to review all the code (excluding the tests).  One thing I would like to discuss is the `by` methods.  I'm in favor of doing it by `Key`.   We might have helper methods `by(org){ return byOrgId(org.getId());`

This way we don't force anyone to go to memcache (or maybe even datastore) to get an entire entity when all we need is the key.

wszarmach commented:
* Corrected the ancestor queries.
* Added Null checks
* Removed silent exception catch
* Removed composite query
* Removed toLower
* Changed Dao method parameters to be Key not Object
* Removed left over line in JasActivityTransformer from manual testing
